### Libraries Installation

In [5]:
# Basic Libraries Installation
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [6]:
# Additional Libraries Installation
import sklearn, scipy, statsmodels, tensorflow
print("Libraries Installed Successfully")

Libraries Installed Successfully


### Steps to Understand the Data 

In [10]:
# Read the file
df = pd.read_excel("Online Retail.xlsx")

In [12]:
# Shows top 5 rows
df.head(5)

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
0,536365,85123A,WHITE HANGING HEART T-LIGHT HOLDER,6,2010-12-01 08:26:00,2.55,17850.0,United Kingdom
1,536365,71053,WHITE METAL LANTERN,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom
2,536365,84406B,CREAM CUPID HEARTS COAT HANGER,8,2010-12-01 08:26:00,2.75,17850.0,United Kingdom
3,536365,84029G,KNITTED UNION FLAG HOT WATER BOTTLE,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom
4,536365,84029E,RED WOOLLY HOTTIE WHITE HEART.,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom


In [14]:
# Shows total no. of rows & columns
df.shape

(541909, 8)

In [15]:
# Shows the name of all the columns
df.columns

Index(['InvoiceNo', 'StockCode', 'Description', 'Quantity', 'InvoiceDate',
       'UnitPrice', 'CustomerID', 'Country'],
      dtype='object')

In [16]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype         
---  ------       --------------   -----         
 0   InvoiceNo    541909 non-null  object        
 1   StockCode    541909 non-null  object        
 2   Description  540455 non-null  object        
 3   Quantity     541909 non-null  int64         
 4   InvoiceDate  541909 non-null  datetime64[ns]
 5   UnitPrice    541909 non-null  float64       
 6   CustomerID   406829 non-null  float64       
 7   Country      541909 non-null  object        
dtypes: datetime64[ns](1), float64(2), int64(1), object(4)
memory usage: 33.1+ MB


### Steps to check the common issues to be identified to clean the data

In [17]:
# Shows the total missing values
df.isnull().sum()

InvoiceNo           0
StockCode           0
Description      1454
Quantity            0
InvoiceDate         0
UnitPrice           0
CustomerID     135080
Country             0
dtype: int64

*From the above output, the column <b>"Description"</b> has <b>1454</b> missing values and the column <b>"CustomerID"</b> has <b>135080</b> missing values.*

In [18]:
# Shows the duplicate rows
df.duplicated().sum()

np.int64(5268)

In [29]:
# Shows the basic stats of the data
df.describe()

,Quantity,InvoiceDate,UnitPrice,CustomerID
count,392692.000000,392692,392692.000000,392692.000000
mean,13.119702,2011-07-10 19:13:07.771892480,3.125914,15287.843865
min,1.000000,2010-12-01 08:26:00,0.001000,12346.000000
25%,2.000000,2011-04-07 11:12:00,1.250000,13955.000000
50%,6.000000,2011-07-31 12:02:00,1.950000,15150.000000
75%,12.000000,2011-10-20 12:53:00,3.750000,16791.000000
max,80995.000000,2011-12-09 12:50:00,8142.750000,18287.000000
std,180.492832,NaN,22.241836,1713.539549


In [30]:
# Check data types
df.dtypes

InvoiceNo              object
StockCode              object
Description            object
Quantity                int64
InvoiceDate    datetime64[ns]
UnitPrice             float64
CustomerID            float64
Country                object
dtype: object

### Cleaning Process of the Data

In [20]:
# Removes the missing values 
df = df.dropna(subset=["Description"])
df = df.dropna(subset=["CustomerID"])

##### *Removing the missing values instead of replacing it with mean or median <br>*

*<b>"CustomerID"</b> is an identifier, not a measurable value so, if we fill it with mean or median would create fake customers <br>*

*<b>"Description"</b> is categorical in nature so, replacing it with mean or median  could fill it with [Unknown] which adds noise and does not help with K-means clustering*

In [21]:
# Removes duplicate rows
df = df.drop_duplicates()

In [31]:
# Convert the data type of (CustomerID) from float to int
df["CustomerID"] = df["CustomerID"].astype(int)

In [25]:
# Review the total no. of rows and columns after removing missing values and duplicate rows
df.shape

(401604, 8)

In [22]:
# Re-check if there are any missing values
df.isnull().sum()

InvoiceNo      0
StockCode      0
Description    0
Quantity       0
InvoiceDate    0
UnitPrice      0
CustomerID     0
Country        0
dtype: int64

In [23]:
# Re-check if there are any duplicate rows
df.duplicated().sum()

np.int64(0)

In [33]:
# Re-check the datatypes
df.dtypes

InvoiceNo              object
StockCode              object
Description            object
Quantity                int64
InvoiceDate    datetime64[ns]
UnitPrice             float64
CustomerID              int64
Country                object
dtype: object

In [24]:
# Checks if dataset contains negative values in (Quantity)
df[df["Quantity"] <= 0]

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
141,C536379,D,Discount,-1,2010-12-01 09:41:00,27.50,14527.0,United Kingdom
154,C536383,35004C,SET OF 3 COLOURED FLYING DUCKS,-1,2010-12-01 09:49:00,4.65,15311.0,United Kingdom
235,C536391,22556,PLASTERS IN TIN CIRCUS PARADE,-12,2010-12-01 10:24:00,1.65,17548.0,United Kingdom
236,C536391,21984,PACK OF 12 PINK PAISLEY TISSUES,-24,2010-12-01 10:24:00,0.29,17548.0,United Kingdom
237,C536391,21983,PACK OF 12 BLUE PAISLEY TISSUES,-24,2010-12-01 10:24:00,0.29,17548.0,United Kingdom
...,...,...,...,...,...,...,...,...
540449,C581490,23144,ZINC T-LIGHT HOLDER STARS SMALL,-11,2011-12-09 09:57:00,0.83,14397.0,United Kingdom
541541,C581499,M,Manual,-1,2011-12-09 10:28:00,224.69,15498.0,United Kingdom
541715,C581568,21258,VICTORIAN SEWING BOX LARGE,-5,2011-12-09 11:57:00,10.95,15311.0,United Kingdom
541716,C581569,84978,HANGING HEART JAR T-LIGHT HOLDER,-1,2011-12-09 11:58:00,1.25,17315.0,United Kingdom


In [26]:
# Dataset returns negative values in (Quantity)
# We have to remove it
df = df[df["Quantity"] >0]

In [27]:
# Check (UnitPrice) if it has any invalid prices
df[df["UnitPrice"] <=0]

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
9302,537197,22841,ROUND CAKE TIN VINTAGE GREEN,1,2010-12-05 14:02:00,0.0,12647.0,Germany
33576,539263,22580,ADVENT CALENDAR GINGHAM SACK,4,2010-12-16 14:36:00,0.0,16560.0,United Kingdom
40089,539722,22423,REGENCY CAKESTAND 3 TIER,10,2010-12-21 13:45:00,0.0,14911.0,EIRE
47068,540372,22090,PAPER BUNTING RETROSPOT,24,2011-01-06 16:41:00,0.0,13081.0,United Kingdom
47070,540372,22553,PLASTERS IN TIN SKULLS,24,2011-01-06 16:41:00,0.0,13081.0,United Kingdom
56674,541109,22168,ORGANISER WOOD ANTIQUE WHITE,1,2011-01-13 15:10:00,0.0,15107.0,United Kingdom
86789,543599,84535B,FAIRY CAKES NOTEBOOK A6 SIZE,16,2011-02-10 13:08:00,0.0,17560.0,United Kingdom
130188,547417,22062,CERAMIC BOWL WITH LOVE HEART DESIGN,36,2011-03-23 10:25:00,0.0,13239.0,United Kingdom
139453,548318,22055,MINI CAKE STAND HANGING STRAWBERY,5,2011-03-30 12:45:00,0.0,13113.0,United Kingdom
145208,548871,22162,HEART GARLAND RUSTIC PADDED,2,2011-04-04 14:42:00,0.0,14410.0,United Kingdom


In [28]:
# Removes the invalid prices
df = df[df["UnitPrice"] > 0]

### Feature Engineering

In [35]:
# Add a new column (TotalPrice)
df["TotalPrice"] = df["Quantity"] * df["UnitPrice"]

In [36]:
# Re-check shape of the dataset to ensure if a new column has been added
df.shape

(392692, 9)

In [37]:
# Snapshot Date
snapshot_date = df["InvoiceDate"].max()

In [38]:
# Create an RFM (Recency, Frequency, Monetary)
rfm = df.groupby("CustomerID").agg({
    "InvoiceDate" : "max",
    "InvoiceNo" : "nunique",
    "TotalPrice" : "sum"
})

In [41]:
# Rename the columns
rfm.columns = ["LastPurchase", "Frequency", "Monetary"]

In [43]:
# Create Recency separately
rfm["Recency"] = (snapshot_date - rfm["LastPurchase"]).dt.days

*Recency was created separately for easier understanding*

In [44]:
# Drop the helper column
rfm = rfm.drop(columns=["LastPurchase"])

In [45]:
rfm.head()

,Frequency,Monetary,Recency
CustomerID,,,
12346,1,77183.60,325
12347,7,4310.00,1
12348,4,1797.24,74
12349,1,1757.55,18
12350,1,334.40,309


#### Scaling

In [46]:
from sklearn.preprocessing import StandardScaler

In [48]:
scaler = StandardScaler()
rfm_scaled = scaler.fit_transform(rfm)

In [49]:
# Convert from Series to DataFrame
rfm_scaled = pd.DataFrame(rfm_scaled,
                          columns=rfm.columns,
                          index=rfm.index)
rfm_scaled.head()

,Frequency,Monetary,Recency
CustomerID,,,
12346,-0.425097,8.363010,2.334574
12347,0.354417,0.251699,-0.905340
12348,-0.035340,-0.027988,-0.175360
12349,-0.425097,-0.032406,-0.735345
12350,-0.425097,-0.190812,2.174578


In [50]:
rfm_scaled.describe()

,Frequency,Monetary,Recency
count,4.338000e+03,4.338000e+03,4.338000e+03
mean,1.801745e-17,2.293130e-17,2.702618e-17
std,1.000115e+00,1.000115e+00,1.000115e+00
min,-4.250965e-01,-2.276151e-01,-9.153401e-01
25%,-4.250965e-01,-1.939190e-01,-7.453445e-01
50%,-2.951776e-01,-1.536162e-01,-4.153533e-01
75%,9.457903e-02,-4.319704e-02,4.946227e-01
max,2.659803e+01,3.096074e+01,2.814561e+00
